### PR TITLE
feat: refactor resource details page

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

This PR refactors the details page of a resource 

### Screenshot/screencast of this PR

details-page

### What issues does this PR fix or reference?

it is part of #1337 

### How to test this PR?

1. open the settings -> resources page 
2. visit the details page of a resource

### CODE REVIEW

**Comment by lstocchi** [comment]
@benoitf still not ready to be reviewed. I opened this for a discussion and to have the code when opening new PRs.
First, should i allow users to edit the values? In the mockup i see. So after a change i save it? Should they be made when the engine is stopped, right? How do i know if something is not editable? I think we miss some info
image
 
Second, I guess that by opening the console tab i should connect to the machine. Should we customize it as we do for the start/stop/delete actions so each provider can list it as a possible action, no?

Maybe for the first iteration i would skip both, wdyt?

**Comment by benoitf** [comment]
to your first question yes it should be possible to edit.  For example I can raise the CPU and memory allocated to my machine. It's just that it needs a restart of the machine ( podman extension should handle that)
First iteration = not editable like today ( or just before you introduced the resource page)

to the second question, we need enhancement on the provider side to enable that. Maybe a Terminal view and a way to register more actions ( than start stop delete)

**Comment by benoitf** [comment]
and yes for the first iteration, skip both

**Review by jeffmaury — COMMENTED** [review]
The log tab is always empty

**Comment by lstocchi** [comment]
> The log tab is always empty

There is no a log command for a machine to retrieve old logs. You see something only when you start/stop a machine and you have the tab opened or you have opened it before

**Comment by lstocchi** [comment]
@jeffmaury updated the PR so that if any start/stop action is executed from PD you can read the logs. There is still a limitation though. If a machine is started outside PD there is no way to retrieve them.

**Comment by lstocchi** [comment]
@benoitf @cdrage @jeffmaury can some of you please give this a look? It's a bit old now.. I spend more time rebasing it that working on other stuff lol. Thank you!!

**Comment by benoitf** [comment]
should we have breadcrumb like in container details page ?
(top left corner)
(could be probably follow-up PR)
image


When clicking on the details of a podman machine, I'm always getting to the details of the same machine

https://github.com/containers/podman-desktop/assets/436777/624c528b-8ee0-40fb-ba4a-fc5dd8ad2989


If I click on start inside kind provider, I don't have any log but terminal is scrolling and scrolling and scrolling but I don't see anything

https://github.com/containers/podman-desktop/assets/436777/f60610fb-6a7d-4f25-b049-8102cc46650d

**Comment by lstocchi** [comment]
@benoitf fixed all your issues 🤞 
@jeffmaury ready to be reviewed again

File: packages/extension-api/src/extension-api.d.ts [hunk 1]
```diff
@@ -308,7 +308,7 @@ declare module '@podman-desktop/api' {
    * It can be turned off/on by the user.
    */
   export interface ProviderAutostart {
-    start(logger: Logger): Promise<void>;
+    start(logger?: Logger): Promise<void>;
```
Comment (jeffmaury) [inline packages/extension-api/src/extension-api.d.ts]: What is the rationale for this change ? Are there some new cases where the logger was not being passed ?

File: packages/main/src/plugin/provider-registry.ts [hunk 2]
```diff
@@ -172,6 +174,7 @@ export class ProviderRegistry {
     const providerImpl = new ProviderImpl(id, providerOptions, this, this.containerRegistry);
     this.count++;
     this.providers.set(id, providerImpl);
+    this.providerIdInternalIdMap.set(providerOptions.id, id);
```
Comment (jeffmaury) [inline packages/main/src/plugin/provider-registry.ts]: Seems to me a provider now has 2 differents ids, can you elaborate

**Review by jeffmaury — COMMENTED** [review]
I also noticed a small discrepency: when you create a resource, the log is empty but if the resource is already created and you start it (eg Kind cluster) then you can see the log.

File: packages/extension-api/src/extension-api.d.ts (hunk 1 above)
Comment (lstocchi) [reply packages/extension-api/src/extension-api.d.ts]: removed. I guess it was some old thing i forgot to remove

File: packages/main/src/plugin/provider-registry.ts (hunk 2 above)
Comment (lstocchi) [reply packages/main/src/plugin/provider-registry.ts]: Actually the provider already has 2 ids, the internal id which is created when the provider is registered (a simple count - first registration 0, then 1, 2 ....) and the provider id.

All calls made by the client use the internalId but an extension that wants to retrieve some info about itself does not have any knowledge about its assigned internal id so it uses its id. 
BTW i removed the map as not really necessary. I use the already existing providers map to find the internal id from the id and call the existing functions.

**Comment by lstocchi** [comment]
> I also noticed a small discrepency: when you create a resource, the log is empty but if the resource is already created and you start it (eg Kind cluster) then you can see the log.

The creation logs is not supported here, this is why you don't see anything. Only start/stop actions. Should the creation log be visible as well? If it fails there will be no details page, otherwise there will be only logs about a successfully creation. Is it required? if so, I'd rather work on it in another PR, it'll be easier to review/fix

**Review by benoitf — COMMENTED** [review]
do you think we should split this PR in multiple PRs ?

I see changes in kind extension, podman extension, extension API, provider registry, custom icon

so it looks like resource page should include only changes on preferences/** , all other changes could be merged separately

File: packages/renderer/src/lib/preferences/Util.spec.ts [hunk 3]
```diff
@@ -37,13 +37,13 @@ test('write array object', () => {
 test('write invalid object', () => {
   writeToTerminal(xtermMock, {} as unknown as string[], 'test');
   // no error reported
-  expect(xtermMock.write).toBeCalledWith(expect.stringContaining('test'));
+  expect(xtermMock.write).not.toBeCalledWith(expect.stringContaining('test'));
```
Comment (benoitf) [inline packages/renderer/src/lib/preferences/Util.spec.ts]: I don't see the rational here ?

File: packages/renderer/src/lib/preferences/Util.spec.ts [hunk 4]
```diff
@@ -37,13 +37,13 @@ test('write array object', () => {
...
 });
 
 test('write undefined object', () => {
   writeToTerminal(xtermMock, undefined as unknown as string[], 'test');
   // no error reported
-  expect(xtermMock.write).toBeCalledWith(expect.stringContaining('test'));
+  expect(xtermMock.write).not.toBeCalledWith(expect.stringContaining('test'));
```
Comment (benoitf) [inline packages/renderer/src/lib/preferences/Util.spec.ts]: I would expect that the test check that something is written

File: packages/main/src/plugin/provider-registry.ts [hunk 5]
```diff
@@ -646,15 +646,24 @@ export class ProviderRegistry {
...
 
   // helper method
-  protected getMatchingProvider(providerId: string): ProviderImpl {
+  protected getMatchingProvider(internalId: string): ProviderImpl {
     // need to find the provider
-    const provider = this.providers.get(providerId);
+    const provider = this.providers.get(internalId);
     if (!provider) {
-      throw new Error(`no provider matching provider id ${providerId}`);
+      throw new Error(`no provider matching provider id ${internalId}`);
```
Comment (benoitf) [inline packages/main/src/plugin/provider-registry.ts]: AFAIK we should not include such changes as part of introducing a new page. It's just a replacement of variable names

File: packages/renderer/src/lib/preferences/Util.ts [hunk 6]
```diff
@@ -37,16 +44,24 @@ export interface IConnectionRestart {
...
+function writeArrayToTerminal(xterm: any, data: any[], colorPrefix: string) {
+  for (const content of data) {
+    if (Array.isArray(content)) {
+      writeArrayToTerminal(xterm, content, colorPrefix);
+    } else if (typeof content === 'string') {
+      writeMultilineString(xterm, content, colorPrefix);
+    }
+  }
+}
+
```
Comment (benoitf) [inline packages/renderer/src/lib/preferences/Util.ts]: I would expect a separate PR for this change, including more tests due to the change of behavior

File: packages/renderer/src/lib/preferences/Util.spec.ts (hunk 3 above)
Comment (lstocchi) [reply packages/renderer/src/lib/preferences/Util.spec.ts]: it is trying to write on the terminal an invalid value, so it should not call the `xterm.write` at all.
Now there is a check that only call `xterm.write` when it deals with a valid string array/string value.
No?

File: packages/main/src/plugin/provider-registry.ts (hunk 5 above)
Comment (lstocchi) [reply packages/main/src/plugin/provider-registry.ts]: For me it clarifies the expected param. The provider has both an internalid and an id. By reading `providerId` i am not sure which is the id it actually represent. Should this be in another PR?

File: packages/renderer/src/lib/preferences/Util.ts (hunk 6 above)
Comment (lstocchi) [reply packages/renderer/src/lib/preferences/Util.ts]: Ok, moving it

**Comment by lstocchi** [comment]
> do you think we should split this PR in multiple PRs ?
> 
> I see changes in kind extension, podman extension, extension API, provider registry, custom icon
> 
> so it looks like resource page should include only changes on preferences/** , all other changes could be merged separately

Ok, i'll create multiple PRs

File: packages/main/src/plugin/provider-registry.ts (hunk 5 above)
Comment (benoitf) [reply packages/main/src/plugin/provider-registry.ts]: I mean, your change is legitimate but be moved out of this PR as it's not related at all to resource details page

**Comment by benoitf** [comment]
thanks, I think it will reduce the scope of each PR and merged it faster

**Comment by benoitf** [comment]
I understand the way of changing lot of stuff everywhere, but contributing it back is lot easier for review/separate of concerns/changes to target a specific changeset

**Comment by lstocchi** [comment]
@benoitf i guess this should be good for a review

**Review by benoitf — APPROVED** [review]
thanks @lstocchi
